http/write_handler.go

```diff
@@ -369,3 +395,37 @@ func compressWithGzip(data io.Reader) (io.Reader, error) {
 
 	return pr, err
 }
+
+type limitedReader struct {
+	*io.LimitedReader
+	err   error
+	close func() error
+}
+
+func newLimitedReadCloser(r io.ReadCloser, n int64) *limitedReader {
+	// read up to max + 1 as limited reader just returns EOF when the limit is reached
+	// or when there is nothing left to read. If we exceed the max batch size by one
+	// then we know the limit has been passed.
+	return &limitedReader{
+		LimitedReader: &io.LimitedReader{R: r, N: n + 1},
+		close:         r.Close,
+	}
+}
+
+// Close returns an ErrMaxBatchSizeExceeded when the wrapped reader
+// exceeds the set limit for number of bytes.
+// This is safe to call more than once as long as the provided io.ReadCloser is
+// also safe to close more than once.
+func (l *limitedReader) Close() (err error) {
+	defer func() {
+		if cerr := l.close(); cerr != nil && err == nil {
+			err = cerr
+		}
```

Comment:
note to self: I could swap the close after the first use for a noop one so it ensures l.close() only called once.